index.js

```diff
@@ -176,67 +176,72 @@ module.exports = class BlindPeering {
 
     for (const mirrorKey of mirrors) {
       const ref = this._getBlindPeer(mirrorKey)
-      const cores = new Set()
 
       base.core.on('migrate', () => {
-        this._mirrorBaseBackground(ref, base, cores, all)
+        this._mirrorBaseBackground(ref, base, all)
       })
 
       base.on('writer', (writer) => {
-        if (!isStaticCore(writer.core) && !all) return
-        this._mirrorBaseWriterBackground(ref, base, writer.core)
+        const always = isStaticCore(writer.core) || all
+        this._mirrorBaseWriterBackground(ref, base, writer, always)
       })
 
       base.on('close', () => {
         this.mirroring.delete(base)
         this._releaseMirror(ref)
-
-        for (const core of cores) {
-          ref.cores.delete(core)
-        }
       })
 
-      promises.push(this._mirrorBaseBackground(ref, base, cores, all))
+      promises.push(this._mirrorBaseBackground(ref, base, all))
     }
 
     return Promise.all(promises)
   }
 
-  async _mirrorBaseWriterBackground (ref, base, core) {
+  async _mirrorBaseWriter (ref, base, core, always) {
+    if (ref.cores.has(core)) return
+
     ref.refs++
-    ref.cores.add(core)
 
-    const referrer = base.wakeupCapability.key
+    try {
+      if (!always && await ref.isReplicating(core)) {
+        return
+      }
+
+      ref.cores.add(core)
+
+      core.on('close', () => {
+        ref.cores.delete(core)
+      })
+
+      const referrer = base.wakeupCapability.key
+      if (this.passive) {
+        await ref.peer.connect()
+      } else {
+        await ref.peer.addCore(core.key, { announce: false, referrer, priority: 1 })
+      }
+    } finally {
+      this._releaseMirror(ref)
+    }
+  }
 
+  async _mirrorBaseWriterBackground (ref, base, core) {
     try {
-      return this.passive
-        ? await ref.peer.connect()
-        : await ref.peer.addCore(core.key, { announce: false, referrer, priority: 1 })
+      await this._mirrorBaseWriter(ref, base, core)
     } catch (e) {
       safetyCatch(e)
-      // ignore
-    } finally {
-      this._releaseMirror(ref)
     }
   }
 
-  _addBaseCores (ref, base, cores, all) {
+  _addBaseCores (ref, base, all) {
     if (this.passive) return Promise.all([ref.peer.connect()])
 
-    const referrer = base.wakeupCapability.key
     const promises = []
 
-    promises.push(ref.peer.addCore(base.local.key, { announce: false, referrer, priority: 1 }))
-
-    ref.cores.add(base.local)
-    cores.add(base.local)
+    promises.push(this._mirrorBaseWriter(ref, base, base.local, true))
 
     for (const writer of base.activeWriters) {
-      if (!isStaticCore(writer.core) && !all) continue
-      promises.push(ref.peer.addCore(writer.core.key, { announce: false, referrer, priority: 1 }))
-
-      ref.cores.add(writer.core)
-      cores.add(writer.core)
+      const always = isStaticCore(writer.core) || all
+      promises.push(this._mirrorBaseWriter(ref, base, writer, always))
```

Comment:
Needs to be `writer.core` (or `mirrorBaseWriter` needs to expect a Writer object)